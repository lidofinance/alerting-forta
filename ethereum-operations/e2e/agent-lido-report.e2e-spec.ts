import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-lido-report e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/lido-report/agent-lido-report")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with Lido Oracle report",
    async () => {
      // todo: should be uncommented after v2
      // const findings = await runTransaction(
      //   "0xe949652989ceed222ad1d1a903f7c925d64a7227b6a286451f9a454f753e9241"
      // );
      // expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with decreased Lido Beacon rewards",
    async () => {
      // todo: should be uncommented after v2
      // const findings = await runTransaction(
      //   "0x1a5eed94c2da9da1ab5d40b723f92c43ce8a06e00b0a369d15561618115ef199"
      // );
      // expect(findings.at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with Lido rebase digest",
    async () => {
      const findings = await runTransaction(
        "0xbf52777e4dd583d52104be96f7da420be977faeee97cc25a89d6c81fa919056f"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
