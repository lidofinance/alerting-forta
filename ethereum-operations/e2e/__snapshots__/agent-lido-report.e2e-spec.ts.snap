// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-lido-report e2e tests should process tx with Lido rebase digest 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-REPORT-REBASE-DIGEST",
  "description": "*APR stats*
APR: 5.26%
Total shares: 5646066.679 -> 5644434.777 1e18 (-0.03%)
Total pooled ether: 6349026.643 -> 6348106.239 ETH (-0.01%)
Time elapsed: 24 hrs 0 sec

*Rewards*
CL rewards: 623.374 (+0.248) ETH
EL rewards: 392.917 (+51.693) ETH
Total: 1016.291 (+51.940) ETH

*Validators*
Count: 196087 (0 newly appeared)

*Withdrawn from*
Withdrawal Vault: 756.285 ETH
EL Vault: 392.917 ETH
Buffer: 787.493 ETH

*Requests finalization*
Finalized: 41
Ether: 1936.695 ETH
Share rate: 1.12467

*Shares*
Burnt: 1722.266 1e18",
  "labels": Array [],
  "metadata": Object {},
  "name": "ℹ️ Lido Report: rebase digest",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;
